### ISSUE TITLE
Buffering wp-admin causes slow page delivery and out of memory errors

### ISSUE THREAD

**Issue by kkmuffme** [woothemes/woocommerce#51890]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

Caused by [https://github.com/woocommerce/woocommerce/blame/trunk/plugins/woocommerce/includes/admin/class-wc-admin.php#L25](<https://github.com/woocommerce/woocommerce/blame/trunk/plugins/woocommerce/includes/admin/class-wc-admin.php#L25>)

WooCommerce output buffers all wp-admin pages to workaround bad code organization on your end, to allow sending headers/doing redirects late.<br>However, this is an extremely bad practice because:<br>a) it increases PHP memory consumption massively, since the whole page is buffered in memory. Especially in wp-admin where pages can be large, this is a problem

b) perceived page loading time in wp-admin is extremely slow. Since the whole page needs to be loaded, before parts can be sent. Usually PHP will send chunks (assuming HTTP 1.1 chunked encoding or by default in HTTP/2 or HTTP/3), which lead to much faster perceived load time and also faster page usability. e.g. simplified: the header gets sent, then parts of the body as the become available. Any slow queries or code will not stop the whole page from loading. Especially any external JS/CSS in header can be loaded early.

Remove the ob_start buffering and correctly fix the issue that prompted that PR to add the output buffering.

### Expected behavior

No buffering of whole output

### Actual behavior

Buffering of output

### Steps to reproduce

in theme functions.php add:

```
add_action( 'in_admin_header', function () {
flush();
sleep( 10 );
}, 10000, 0 );
```

Load /wp-admin/<br>It will take at least 10 seconds before you see anything.

Now comment `WC_Admin` line referenced above:<br>`add_action( 'admin_init', array( $this, 'buffer' ), 1 );`

Load wp-admin again.<br>You will see the first parts of wp-admin (e.g. header) almost instantly, while the rest of the page takes 10 seconds to load (because of the sleep, which is just there to visualize the issue)

### WordPress Environment

WC 9.4

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](<https://woocommerce.com/products/storefront/>).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by kkmuffme** [comment]
And to add insult to injury, there's no simple way to remove that action, since the class instance isn't available anywhere.
So you have to manually filter through the filters in a filter to remove that action 👎

**Comment by waclawjacek** [comment]
Hi! Thanks for the report. I am not able to reproduce this. What browser are you using?

**Comment by kkmuffme** [comment]
This has nothing to do with the browser used.
Most likely your web server uses a buffer too

Did you see the first part of the page in both cases after the same amount of time? Or with 10 seconds delay when you add the reproduction code.

**Comment by waclawjacek** [comment]
No, everything was loading until the entire page contents were ready and then the page rerendered.

I'll try with buffering disabled on my machine and let you know.

**Comment by kkmuffme** [comment]
Yeah, then it was buffered by your web server - if nginx you can just set `header( 'X-Accel-Buffering: no' );` to disable nginx's internal buffering for the request

Btw if your test site is behind Cloudflare, there is a bug in Cloudflare atm if it encodes with zstd it causes the same behavior https://community.cloudflare.com/t/zstd-prevents-text-html-streaming-output-free-plan/701335/24

I'm unable to provide further support for free, if someone who has experience with buffering could please take a look at this

**Comment by pmcpinto** [comment]
Thanks for such a detailed report, it's really helpful.

I know this has been open a while, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

I'd like to get an engineer's eyes on this before we make a call, so I've flagged it for developer feedback. Someone will take a proper look at the repro steps and how it's behaving now.
